File: AutomaticDrawing.Recognition.Clustering/Clustering/Hierarchy/GridHierarchy/GridHierarchy.vb
Imports System.Numerics
Imports AutomaticDrawing.Core
Imports AutomaticDrawing.Recognition.Clustering
''' <summary>
''' 由单元格表示的层
''' </summary>
Public Class GridHierarchy
    Inherits HierarchyBase
    ''' <summary>
    ''' 单元格
    ''' </summary>
    Public Property Grid As Cluster(,)

    Private Shared OffsetX() As Integer = {-1, 0, 1, 1, 1, 0, -1, -1}
    Private Shared OffsetY() As Integer = {-1, -1, -1, 0, 1, 1, 1, 0}

    ''' <summary>
    ''' 创建并初始化一个实例
    ''' </summary>
    Public Sub New(w As Integer, h As Integer)
        ReDim Grid(w - 1, h - 1)
    End Sub

    ''' <summary>
    ''' 由指定的<see cref="PixelData"/>对象创建一个实例
    ''' </summary>
    Public Shared Function CreateFromPixels(pixels As PixelData) As GridHierarchy
        Dim result As New GridHierarchy(pixels.Width, pixels.Height)
        For i = 0 To pixels.Width - 1
            For j = 0 To pixels.Height - 1
                Dim cluster As New Cluster With
                {
                    .Position = New Vector2(i, j),
                    .Color = pixels.Colors(i, j)
                }
                result.Grid(i, j) = cluster
                result.AddCluster(cluster)
            Next
        Next
        Return result
    End Function

    Public Overrides Function Generate() As IHierarchy
        Dim result As New GroupHierarchy With {.Rank = Me.Rank + 1}

        For Each SubCluster In Clusters
            Dim similar As Cluster = SubCluster.GetMostSimilar(GetNeighbours(SubCluster))
            If similar IsNot Nothing Then
                result.AddCluster(Cluster.Combine(SubCluster, similar))
            End If
        Next

        Return result
    End Function

    Private Function GetNeighbours(cluster As Cluster) As List(Of Cluster)
        Dim result As New List(Of Cluster)
        Dim xBound As Integer = Grid.GetUpperBound(0)
        Dim yBound As Integer = Grid.GetUpperBound(1)
        Dim dx, dy As Integer
        Dim x As Integer = cluster.Position.X
        Dim y As Integer = cluster.Position.Y
        For i = 0 To 7
            dx = x + OffsetX(i)
            dy = y + OffsetY(i)
            If (dx >= 0 AndAlso dy >= 0 AndAlso dx <= xBound AndAlso dy <= yBound) Then
                result.Add(Grid(dx, dy))
            Else
                Continue For
            End If
        Next
        Return result
    End Function
End Class

File: AutoPaint.Recognition.Clustering/Clustering/Hierarchy/GridHierarchy.cs
﻿using AutoPaint.Core;
using AutoPaint.Utilities;
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;

namespace AutoPaint.Recognition.Clustering
{
    public class GridHierarchy : HierarchyBase
    {
        public ConcurrentBag<Cluster>[,] Grid { get; set; }

        public int Width { get; set; }
        public int Height { get; set; }
        public float Size { get; set; }

        private static readonly int[] OffsetX9 = new[] { 0, -1, 1, 0, 0, -1, 1, -1, 1 };
        private static readonly int[] OffsetY9 = new[] { 0, 0, 0, -1, 1, -1, 1, 1, -1 };

        private static readonly int[] OffsetX4 = new[] { 0, 1, 0, -1 };
        private static readonly int[] OffsetY4 = new[] { -1, 0, 1, 0 };

        public GridHierarchy(int w, int h, float size)
        {
            if (w < 1)
            {
                w = 1;
            }

            if (h < 1)
            {
                h = 1;
            }

            this.Width = w;
            this.Height = h;
            this.Size = size;

            Grid = new ConcurrentBag<Cluster>[w - 1 + 1, h - 1 + 1];
            for (var i = 0; i <= w - 1; i++)
            {
                for (var j = 0; j <= h - 1; j++)
                    Grid[i, j] = new ConcurrentBag<Cluster>();
            }
        }

        public static GridHierarchy CreateFromPixels(PixelData pixels)
        {
            GridHierarchy result = new GridHierarchy(pixels.Width, pixels.Height, 1);
            for (var i = 0; i <= pixels.Width - 1; i++)
            {
                for (var j = 0; j <= pixels.Height - 1; j++)
                {
                    Cluster cluster = new Cluster()
                    {
                        Position = new Vector2(i, j),
                        Color = pixels.Colors[i, j]
                    };
                    result.Grid[i, j].Add(cluster);
                    result.Clusters.Add(cluster);
                }
            }

            return result;
        }

        public override IHierarchy Generate()
        {
            float rate = 2.0F;
            float newSize = this.Size * rate;
            GridHierarchy result = new GridHierarchy(System.Convert.ToInt32(Math.Ceiling(this.Width / (double)rate) + 1), System.Convert.ToInt32(Math.Ceiling(this.Height / (double)rate) + 1), newSize) { Rank = this.Rank + 1 };

            // Combine to a new cluster
            //foreach (var cluster in Clusters)
            //{
            //    Cluster similar = cluster.GetMostSimilar(GetNeighbours(cluster)).FirstOrDefault();
            //    if (similar != null)
            //    {
            //        if (cluster.Parent == null && similar.Parent == null)
            //            result.Clusters.Add(Cluster.Combine(cluster, similar));
            //        else
            //            Cluster.Combine(cluster, similar);
            //    }
            //    else
            //    {
            //        result.Clusters.Add(cluster);
            //    }
            //}

            Parallel.ForEach(Clusters, cluster =>
            {
                Cluster similar = cluster.GetMostSimilar(GetNeighbours(cluster)).FirstOrDefault();
                if (similar != null)
                {
                    if (cluster.Parent == null && similar.Parent == null)
                        result.Clusters.Add(Cluster.Combine(cluster, similar));
                    else
                        Cluster.Combine(cluster, similar);
                }
                else
                {
                    result.Clusters.Add(cluster);
                }
            });

            // Initialize properties
            Parallel.ForEach(result.Clusters, cluster =>
            {
                cluster.Position = cluster.GetAveragePosition();
                cluster.Color = cluster.GetAverageColor();
            });

            // Locate to cell
            Parallel.ForEach(result.Clusters, cluster =>
            {
                Vector2 p = cluster.Position;
                int x = System.Convert.ToInt32(p.X / (double)result.Size);
                int y = System.Convert.ToInt32(p.Y / (double)result.Size);
                result.Grid[x, y].Add(cluster);
            });

            // Calculate edge vector
            Parallel.ForEach(result.Clusters, cluster =>
            {
                cluster.EdgeVector = result.GetNeighboursEdgeVector(cluster); ;
            });

            return result;
        }

        public override string ToString()
        {
            return $"Rank:{Rank}Clusters.Count:{Clusters.Count}";
        }

        public List<Cluster> GetNeighbours(Cluster cluster)
        {
            List<Cluster> result = new List<Cluster>(9);
            int xBound = Grid.GetUpperBound(0);
            int yBound = Grid.GetUpperBound(1);
            int dx, dy;
            int x = System.Convert.ToInt32(cluster.Position.X / (double)Size);
            int y = System.Convert.ToInt32(cluster.Position.Y / (double)Size);
            for (var i = 0; i < 9; i++)
            {
                dx = x + OffsetX9[i];
                dy = y + OffsetY9[i];
                if ((dx >= 0 && dy >= 0 && dx <= xBound && dy <= yBound))
                    result.AddRange(Grid[dx, dy]);
                else
                    continue;
            }
            result.Remove(cluster);
            return result;
        }

        public Vector4 GetNeighboursEdgeVector(Cluster cluster)
        {
            float[] floats = new float[4];
            int xBound = Grid.GetUpperBound(0);
            int yBound = Grid.GetUpperBound(1);
            int dx, dy;
            int x = System.Convert.ToInt32(cluster.Position.X / (double)Size);
            int y = System.Convert.ToInt32(cluster.Position.Y / (double)Size);

            for (var i = 0; i < 4; i++)
            {
                dx = x + OffsetX4[i];
                dy = y + OffsetY4[i];
                if ((dx >= 0 && dy >= 0 && dx <= xBound && dy <= yBound))
                {
                    if (Grid[dx, dy].Count == 0)
                    {
                        floats[i] = 0.0f;
                    }
                    else
                    {
                        var neighbourColor = ColorHelper.FindClosestColor(Grid[dx, dy].Select(v => v.Color));
                        floats[i] = ColorHelper.GetColorSimilarity(cluster.Color, neighbourColor);
                    }
                }
                else
                {
                    floats[i] = 0.0f;
                }
            }

            return new Vector4(floats[0], floats[1], floats[2], floats[3]);
        }
    }
}
